keps/sig-apps/20190716-cronjob-new-concurrency-policy-catchup.md

```diff
@@ -0,0 +1,175 @@
+last-updated: 2019-07-16
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# New Cronjob Concurrency Policy - CatchUp
+
+## Table of Contents
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives](#alternatives)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This document proposes a plan to add a new concurrency policy `CatchUp` to Cronjob,
+such that missed jobs are not skipped and will be scheduled in sequential order.
+Additionally, two derived minor improvements are proposed to better support this feature: 
+(1) Expose the original planned run time to Jobs.
+(2) Only conduct the "maximum 100 missed schedules" check while concurrency policy is set to `Allow`.
+
+## Motivation
+
+Currently if any schedule misses, cronjob controller will skipped it immediately no matter what concurrency policy is in use.
+This behavior, however, does not fit for cronjobs that need to know the exact time range of the data they are going to process.
+For example, I have a cronjob that runs hourly and only processes the logs within the previous hour.
+Say the job now takes 3 hours to run due to a bug, then many jobs will be skipped sporadically,
```

Comment:
In `Allow` mod, everything should be fine, while in `Forbid` mod, my set up is that the job runs every hour and now takes 3 hours to run due to a bug. The result would be:
```
1(run) 2(skip) 3(s) 4(r) 5(s) 6(s) 7(r) 8(s) 9(s) 10(r) 11(s) 12(s) 13(r) 14(s) 15(s) 16(r) 17(s) 18(s) 19(r) 20(s) 21(s) 22(r) 23(s) 24(s)  
```
As you can see jobs will run at 1:00, 4:00, 7:00, 10:00, 13:00, 16:00, 19:00, 22:00, since each one of them were set to only process logs in the previous hour, we would end up missing 16 hours of data. Then user needs to figure out what time slices are missed and backfill them manually. This is what I meant by "skipped sporadically".

Hope it answers the confusion.

In addition, as I described in the next paragraph, this type of cronjob is yet fully supported in Kubernetes at the moment. Because ideally this type of cronjobs need to know when they were suppose to run, so it can figure out what time range of the data it should process.